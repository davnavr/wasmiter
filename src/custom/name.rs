//! Types to parse and describe the structure of the
//! [`name` section, described in the WebAssembly specification appendix](https://webassembly.github.io/spec/core/appendix/custom.html),
//! which associates UTF-8 [`Name`](parser::name::Name)s with definitions in a module.

use crate::{
    bytes::{Bytes, Window},
    index,
    parser::{self, AscendingOrder, ResultExt as _},
    sections::{Section, SectionSequence},
};
use core::fmt::Debug;

mod name_assoc;
mod name_map;

pub use name_assoc::NameAssoc;
pub use name_map::NameMap;

/// Represents a
/// [name subsection](https://webassembly.github.io/spec/core/appendix/custom.html#subsections)
/// within the [`NameSection`].
#[derive(Clone, Copy)]
#[non_exhaustive]
pub enum NameSubsection<B: Bytes> {
    /// The
    /// [*module name subsection*](https://webassembly.github.io/spec/core/appendix/custom.html#module-names)
    /// assigns a name to the WebAssembly module.
    ModuleName(parser::name::Name<B>),
    /// The
    /// [*function name subsection*](https://webassembly.github.io/spec/core/appendix/custom.html#function-names)
    /// assigns names to the functions of a WebAssembly module.
    FunctionName(NameMap<index::FuncIdx, u64, B>),
}

/// Result type used when interpreting the contents of a [`NameSubsection`].
pub type InterpretedNameSubsection<B> =
    Result<parser::Result<NameSubsection<Window<B>>>, Section<B>>;

impl<B: Bytes> NameSubsection<Window<B>> {
    /// Attempts to interpret the contents of the given name subsection.
    ///
    /// Returns `Err(_)` if the section's
    /// [*id*](https://webassembly.github.io/spec/core/appendix/custom.html#subsections) is not
    /// recognized.
    ///
    /// Returns `Ok(Err(_))` if the section **was** recognized, but an attempt to parse a field
    /// failed.
    pub fn interpret(section: Section<B>) -> InterpretedNameSubsection<B> {
        match section.id() {
            0 => {
                let contents = section.into_contents();
                let mut offset = contents.base();
                Ok(parser::name::parse(&mut offset, contents).map(Self::ModuleName))
            }
            1 => {
                let contents = section.into_contents();
                Ok(NameMap::new(contents.base(), contents).map(Self::FunctionName))
            }
            _ => Err(section),
        }
    }
}

impl<B: Bytes> Debug for NameSubsection<B> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::ModuleName(name) => f.debug_tuple("ModuleName").field(name).finish(),
            Self::FunctionName(names) => f.debug_tuple("FunctionName").field(names).finish(),
        }
    }
}

/// Represents the sequence of subsections in the
/// [`name` section](https://webassembly.github.io/spec/core/appendix/custom.html).
///
/// Section IDs are checked to ensure that they are in **ascending** order.
#[derive(Clone, Copy)]
pub struct NameSection<B: Bytes> {
    order: AscendingOrder<u8>,
    first: bool,
    sections: SectionSequence<B>,
}

impl<B: Bytes> NameSection<B> {
    /// Creates a [`NameSection`] from the given sequence of `sections`.
    pub fn new(sections: SectionSequence<B>) -> Self {
        Self {
            order: AscendingOrder::new(),
            first: true,
            sections,
        }
    }

    /// Consumes the [`NameSection`], returning the remaining subsections.
    pub fn into_sections(self) -> SectionSequence<B> {
        self.sections
    }

    fn parse_inner<'a, U: Bytes, F>(&'a mut self, f: F) -> Option<InterpretedNameSubsection<U>>
    where
        U: 'a,
        F: FnOnce(Section<&'a B>) -> Section<U>,
    {
        match self.sections.parse().context("name subsection") {
            Ok(Some(section)) => match self
                .order
                .check(section.id(), self.first)
                .context("invalid name subsection order")
            {
                Ok(_) => {
                    self.first = false;
                    Some(NameSubsection::interpret(f(section)))
                }
                Err(e) => Some(Ok(Err(e))),
            },
            Ok(None) => None,
            Err(e) => Some(Ok(Err(e))),
        }
    }

    /// Attempts to parse the next subsection, returning `None` if no more remain.
    ///
    /// If a subsection is not recognized, `Some(Err(_))` is returned.
    ///
    /// # Errors
    ///
    /// `Some(Ok(Err(_)))` is returned for any parser errors or if section IDs are not in **ascending order**.
    pub fn parse(&mut self) -> Option<InterpretedNameSubsection<&B>> {
        self.parse_inner(core::convert::identity)
    }
}

impl<B: Clone + Bytes> Iterator for NameSection<B> {
    type Item = InterpretedNameSubsection<B>;

    fn next(&mut self) -> Option<Self::Item> {
        self.parse_inner(|subsec| subsec.cloned())
    }
}

impl<B: Bytes> Debug for NameSection<B> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        let mut list = f.debug_list();
        for subsec in self.sections.borrowed() {
            let err;
            let unknown_subsection;
            let known_subsection;

            list.entry(match subsec {
                Ok(section) => match NameSubsection::interpret(section) {
                    Ok(Err(e)) => {
                        err = parser::Result::<()>::Err(e);
                        &err
                    }
                    Ok(Ok(known)) => {
                        known_subsection = parser::Result::Ok(known);
                        &known_subsection
                    }
                    Err(unknown) => {
                        unknown_subsection = parser::Result::Ok(unknown);
                        &unknown_subsection
                    }
                },
                Err(e) => {
                    err = parser::Result::<()>::Err(e);
                    &err
                }
            });
        }
        list.finish()
    }
}
