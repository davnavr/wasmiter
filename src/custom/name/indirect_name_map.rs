use crate::{
    custom::name::NameMap,
    index::Index,
    input::{BorrowInput, CloneInput, HasInput, Input},
    parser::{AscendingOrder, Offset, Result, ResultExt as _, Vector},
};

/// An
/// [*indirect name map*](https://webassembly.github.io/spec/core/appendix/custom.html#name-maps)
/// associates [`NameMap`]s with a priamry index.
///
/// Like a [`NameMap`], each primary index is checked to ensure they are unique and in increasing
/// order.
#[derive(Clone, Copy)]
pub struct IndirectNameMap<K: Index, V: Index, O: Offset, I: Input> {
    entries: Vector<O, I>,
    order: AscendingOrder<u32, K>,
    _marker: core::marker::PhantomData<V>,
}

impl<K: Index, V: Index, O: Offset, I: Input> From<Vector<O, I>> for IndirectNameMap<K, V, O, I> {
    fn from(entries: Vector<O, I>) -> Self {
        Self {
            entries,
            order: AscendingOrder::new(),
            _marker: core::marker::PhantomData,
        }
    }
}

impl<K: Index, V: Index, O: Offset, I: Input> IndirectNameMap<K, V, O, I> {
    /// Parses a [`IndirectNameMap`] starting at the given `offset`.
    pub fn new(offset: O, input: I) -> Result<Self> {
        Vector::parse(offset, input).map(Self::from)
    }

    /// Gets the remaining number of pairs in the [`IndirectNameMap`].
    pub fn remaining_count(&self) -> u32 {
        self.entries.remaining_count()
    }

    /// Parses the next primary index and [`NameMap`] pair.
    pub fn parse<T, F>(&mut self, f: F) -> Result<Option<T>>
    where
        F: FnOnce(K, &mut NameMap<V, &mut u64, &I>) -> Result<T>,
    {
        self.entries
            .advance_with_index(|i, offset, input| {
                let primary_index: K = crate::component::index(offset, input)
                    .context("could not parse primary index for pair")?;

                self.order
                    .check(primary_index, i == 0)
                    .context("primary index is not valid")?;

                let mut name_map =
                    NameMap::new(offset, input).context("could not parse name map for pair")?;

                let result = f(primary_index, &mut name_map)?;
                name_map.finish()?;
                Result::Ok(result)
            })
            .transpose()
            .context("could not parse entry in indirect name map")
    }
}

impl<K: Index, V: Index, O: Offset, I: Input> HasInput<I> for IndirectNameMap<K, V, O, I> {
    #[inline]
    fn input(&self) -> &I {
        self.entries.input()
    }
}

impl<'a, K: Index, V: Index, O: Offset, I: Input + 'a> BorrowInput<'a, I>
    for IndirectNameMap<K, V, O, I>
{
    type Borrowed = IndirectNameMap<K, V, u64, &'a I>;

    #[inline]
    fn borrow_input(&'a self) -> Self::Borrowed {
        IndirectNameMap {
            entries: self.entries.borrow_input(),
            order: self.order,
            _marker: core::marker::PhantomData,
        }
    }
}

impl<K: Index, V: Index, O: Offset, I: Input> core::fmt::Debug for IndirectNameMap<K, V, O, I> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        struct Entry<K: Index, V: Index, I: Input> {
            key: K,
            names: NameMap<V, u64, I>,
        }

        impl<K: Index, V: Index, I: Input> core::fmt::Debug for Entry<K, V, I> {
            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
                f.debug_struct("Entry")
                    .field("key", &self.key)
                    .field("names", &self.names)
                    .finish()
            }
        }

        let mut entries = self.borrow_input();
        let mut list = f.debug_list();
        loop {
            let result = entries.parse(|key, names| {
                list.entry(&Entry {
                    key,
                    names: names.clone_input(),
                });
                Result::Ok(())
            });

            if let Err(e) = result.as_ref() {
                list.entry(&core::result::Result::<(), _>::Err(e));
            }

            if matches!(result, Ok(None) | Err(_)) {
                break;
            }
        }

        list.finish()
    }
}
