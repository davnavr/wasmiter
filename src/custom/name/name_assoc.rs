use crate::{
    bytes::Bytes,
    index::Index,
    parser::{self, name::Name, ResultExt as _},
};

/// A [**nameassoc**](https://webassembly.github.io/spec/core/appendix/custom.html#name-maps)
/// associates an [`Index`] with a [`Name`].
#[derive(Clone, Copy)]
pub struct NameAssoc<I: Index, B: Bytes> {
    index: I,
    name: Name<B>,
}

impl<I: Index, B: Bytes> NameAssoc<I, B> {
    /// Parses a [`NameAssoc`].
    pub fn parse(offset: &mut u64, bytes: B) -> parser::Result<Self> {
        Ok(Self {
            index: crate::component::index(offset, &bytes).context("index of nameassoc pair")?,
            name: parser::name::parse(offset, bytes).context("name of nameassoc pair")?,
        })
    }

    /// Gets the index.
    #[inline]
    pub fn index(&self) -> I {
        self.index
    }

    /// Gets the name.
    #[inline]
    pub fn name(&self) -> &Name<B> {
        &self.name
    }
}

impl<I: Index, B: Clone + Bytes> NameAssoc<I, &B> {
    pub(super) fn dereferenced(&self) -> NameAssoc<I, B> {
        NameAssoc {
            index: self.index,
            name: self.name.really_cloned(),
        }
    }
}

impl<I: Index, B: Bytes> core::fmt::Debug for NameAssoc<I, B> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("NameAssoc")
            .field("index", &self.index)
            .field("name", &self.name)
            .finish()
    }
}
