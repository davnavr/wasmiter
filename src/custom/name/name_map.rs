use crate::{
    bytes::Bytes,
    component::IndexOrder,
    custom::name::NameAssoc,
    index::Index,
    parser::{Offset, Result, ResultExt as _, Vector},
};

/// A [*name map*](https://webassembly.github.io/spec/core/appendix/custom.html#name-maps)
/// is a sequence of [`NameAssoc`] entries, which associate indices with names.
///
/// Each index is checked in order to ensure they are unique and in increasing order.
#[derive(Clone, Copy)]
pub struct NameMap<I: Index, O: Offset, B: Bytes> {
    order: IndexOrder<I>,
    entries: Vector<O, B>,
}

impl<I: Index, O: Offset, B: Bytes> From<Vector<O, B>> for NameMap<I, O, B> {
    fn from(entries: Vector<O, B>) -> Self {
        Self {
            order: IndexOrder::new(),
            entries,
        }
    }
}

impl<I: Index, O: Offset, B: Bytes> NameMap<I, O, B> {
    /// Creates a new [`NameMap`] to parse at the given `offset`.
    pub fn new(offset: O, bytes: B) -> Result<Self> {
        Vector::parse(offset, bytes).map(Self::from)
    }

    /// Parses the next entry in the [`NameMap`].
    pub fn parse(&mut self) -> Result<Option<NameAssoc<I, &B>>> {
        self.entries
            .advance_with_index(|i, offset, bytes| {
                let name_assoc =
                    NameAssoc::parse(offset, bytes).context("while parsing name map entry")?;

                self.order
                    .check(name_assoc.index(), i == 0)
                    .context("name map index was invalid")?;

                Ok(name_assoc)
            })
            .transpose()
    }

    fn borrowed(&self) -> NameMap<I, u64, &B> {
        NameMap {
            order: self.order,
            entries: self.entries.borrowed(),
        }
    }
}

impl<I: Index, O: Offset, B: Clone + Bytes> Iterator for NameMap<I, O, B> {
    type Item = Result<NameAssoc<I, B>>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.parse() {
            Ok(None) => None,
            Err(e) => Some(Err(e)),
            Ok(Some(name_assoc)) => Some(Ok(name_assoc.dereferenced())),
        }
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.entries.size_hint()
    }
}

impl<I: Index, O: Offset, B: Clone + Bytes> core::iter::FusedIterator for NameMap<I, O, B> {}

impl<I: Index, O: Offset, B: Bytes> core::fmt::Debug for NameMap<I, O, B> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_list().entries(self.borrowed()).finish()
    }
}
