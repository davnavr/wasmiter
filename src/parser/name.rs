//! Types and functions for parsing UTF-8 strings from [`Input`] bytes.

use crate::{
    input::{self, BorrowInput, Input, Window},
    parser::{self, Error, ResultExt as _},
};

mod char_iterators;
mod error;
mod name_fmt;

pub use char_iterators::{Chars, CharsLossy};
pub use error::{InvalidCodePoint, NameError};

/// A UTF-8 string [name](https://webassembly.github.io/spec/core/binary/values.html#names).
#[derive(Clone, Copy)]
pub struct Name<I: Input> {
    input: I,
    offset: u64,
    length: u32,
}

impl<I: Input> Name<I> {
    /// Reads a length-prefixed UTF-8 string from the given [`Input`], starting at the given
    /// `offset`.
    ///
    /// # Error
    ///
    /// Returns an error if the length could not be parsed.
    pub fn new(input: I, offset: &mut u64) -> parser::Parsed<Self> {
        Ok(Self {
            length: parser::leb128::u32(offset, &input).context("string length")?,
            offset: *offset,
            input,
        })
    }

    /// Gets the offset to the first byte of the UTF-8 string contents.
    #[inline]
    pub fn offset(&self) -> u64 {
        self.offset
    }

    /// Gets the length, in bytes, of the UTF-8 string contents.
    #[inline]
    pub fn length(&self) -> u64 {
        self.length.into()
    }

    /// Returns an iterator over the [`char`]s of the [`Name`], returning a [`NameError`] for
    /// invalid code points or failures retrieving bytes from the [`Input`].
    #[inline]
    pub fn chars(self) -> Chars<I> {
        Chars::new(self)
    }

    /// Returns an iterator over the [`char`]s of the [`Name`], emitting a
    /// [`char::REPLACEMENT_CHARACTER`] for each invalid code point or failure to retrieve
    /// bytes from the [`Input`].
    #[inline]
    pub fn chars_lossy(self) -> CharsLossy<I> {
        CharsLossy::new(Chars::new(self))
    }

    /// Copies the contents of the [`Name`] into the specified `buffer`.
    ///
    /// If the length of the `buffer` is less than the length, in bytes, of the [`Name`], then only
    /// a portion of the [`Name`] contents is copied.
    ///
    /// # Errors
    ///
    /// Returns an error if the name bytes could not be fetched from the [`Input`].
    pub fn copy_to_slice<'b>(&self, buffer: &'b mut [u8]) -> parser::Parsed<&'b mut [u8]> {
        let length = core::cmp::min(
            usize::try_from(self.length).ok().unwrap_or(usize::MAX),
            buffer.len(),
        );

        let destination: &'b mut [u8] = &mut buffer[0..length];

        self.input
            .read_exact_at(self.offset, destination)
            .context("string contents")?;

        Ok(destination)
    }

    /// Returns the contents of the [`Name`] as a [`Window`](input::Window).
    pub fn into_bytes_window(self) -> input::Window<I> {
        let offset = self.offset;
        let length = self.length();
        input::Window::with_offset_and_length(self.input, offset, length)
    }

    /// Attempts to compare this [`Name`] to a [`str`]ing, returning `true` if they are equal.
    ///
    /// # Errors
    ///
    /// Returns an error if the name bytes could not be feteched from the [`Input`].
    #[inline]
    pub fn try_eq_str(&self, s: &str) -> parser::Parsed<bool> {
        self.borrow_input()
            .into_bytes_window()
            .try_eq_at(self.offset, s.as_bytes())
            .map_err(Into::into)
    }
}

impl<I: Input> input::HasInput<I> for Name<I> {
    #[inline]
    fn input(&self) -> &I {
        &self.input
    }
}

impl<'a, I: Input + 'a> BorrowInput<'a, I> for Name<I> {
    type Borrowed = Name<&'a I>;

    #[inline]
    fn borrow_input(&'a self) -> Name<&'a I> {
        Name {
            input: &self.input,
            offset: self.offset,
            length: self.length,
        }
    }
}

impl<'a, I: Clone + Input + 'a> input::CloneInput<'a, I> for Name<&'a I> {
    type Cloned = Name<I>;

    #[inline]
    fn clone_input(&self) -> Name<I> {
        Name {
            input: self.input.clone(),
            offset: self.offset,
            length: self.length,
        }
    }
}

impl<I: Input> From<Name<I>> for Window<I> {
    /// Converts a [`Name`] into a [`Window`] by calling [`Name::into_bytes_window`].
    #[inline]
    fn from(name: Name<I>) -> Self {
        name.into_bytes_window()
    }
}

impl<I: Input> From<Name<input::Window<I>>> for Name<I> {
    fn from(name: Name<input::Window<I>>) -> Name<I> {
        let mut length = core::cmp::min(
            name.length,
            u32::try_from(name.input.length()).unwrap_or(u32::MAX),
        );

        if name.offset > name.input.base() + name.input.length() {
            length = 0;
        }

        Name {
            offset: name.offset,
            length,
            input: name.input.into_inner(),
        }
    }
}

#[cfg(feature = "alloc")]
#[cfg_attr(doc_cfg, doc(cfg(feature = "alloc")))]
impl<I: Input> Name<I> {
    /// Allocates a byte vector to contain the contents of the [`Name`].
    ///
    /// # Error
    ///
    /// Returns an error if the operation to read the characters from the [`Input`] fails.
    pub fn into_bytes(self) -> parser::Parsed<alloc::vec::Vec<u8>> {
        let mut bytes = alloc::vec![0u8; self.length.try_into().unwrap_or(usize::MAX)];
        self.copy_to_slice(&mut bytes)?;
        Ok(bytes)
    }

    /// Allocates a [`String`](alloc::string::String) containing the contents of the [`Name`].
    ///
    /// # Error
    ///
    /// Returns an error if the operation to read the characters from the [`Input`] fails, or if
    /// the [`Name`] is not valid UTF-8.
    pub fn try_into_string(self) -> parser::Parsed<alloc::string::String> {
        alloc::string::String::from_utf8(self.into_bytes()?).map_err(Into::into)
    }
}

/// Parses a UTF-8 string [`Name`].
pub fn parse<I: Input>(offset: &mut u64, input: I) -> parser::Parsed<Name<I>> {
    let name = Name::new(input, offset)?;
    input::increment_offset(offset, name.length() as usize)?;
    Ok(name)
}

impl<'a> TryFrom<&'a [u8]> for Name<&'a [u8]> {
    type Error = Error;

    fn try_from(bytes: &'a [u8]) -> parser::Parsed<Self> {
        let actual_length = bytes.len();
        if let Ok(length) = u32::try_from(actual_length) {
            Ok(Self {
                length,
                input: bytes,
                offset: 0,
            })
        } else {
            #[cold]
            #[inline(never)]
            fn slice_too_large(length: usize) -> Error {
                Error::new(parser::ErrorKind::InvalidFormat).with_context(
                    parser::Context::from_closure(move |f| {
                        write!(f, "byte slice has a length of {length}, which is too large")
                    }),
                )
            }

            Err(slice_too_large(actual_length))
        }
    }
}

/// Interprets the `str` as a [`Name`] originating from a byte slice.
///
/// # Examples
///
/// ```
/// # fn main() -> Result<(), wasmiter::parser::Error> {
/// use wasmiter::parser::name::Name;
/// let s = "hello";
/// let n = Name::try_from(s)?;
/// assert_eq!(format!("{n}"), s);
/// # Ok(())
/// # }
/// ```
impl<'a> TryFrom<&'a str> for Name<&'a [u8]> {
    type Error = Error;

    fn try_from(s: &'a str) -> parser::Parsed<Self> {
        Self::try_from(s.as_bytes())
    }
}

impl<I: Input> IntoIterator for Name<I> {
    type IntoIter = Chars<I>;
    type Item = Result<char, NameError>;

    #[inline]
    fn into_iter(self) -> Self::IntoIter {
        self.chars()
    }
}
