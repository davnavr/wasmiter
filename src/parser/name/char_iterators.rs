use crate::{
    input::{BorrowInput, HasInput, Input},
    parser::name::{InvalidCodePoint, Name, NameError},
};

#[derive(Clone, Copy, Default)]
struct CharsBuffer {
    /// - Lower 4 bits contain number of saved bytes in `buffer`.
    /// - Upper 4 bits contain number of valid UTF-8 bytes in `buffer`.
    lengths: u8,
    /// Length of the error after `valid_len`.
    bad_sequence: Option<core::num::NonZeroU8>,
    /// # Safety
    ///
    /// The first `valid_len` bytes **must** be valid UTF-8.
    buffer: [u8; 15],
}

impl CharsBuffer {
    #[inline]
    fn saved_len(&self) -> u8 {
        self.lengths & 0xF
    }

    #[inline]
    fn valid_len(&self) -> u8 {
        self.lengths >> 4
    }

    fn valid(&self) -> &str {
        let valid_bytes = &self.buffer[0..usize::from(self.valid_len())];
        if cfg!(debug_assertions) {
            core::str::from_utf8(valid_bytes).unwrap()
        } else {
            // Safety: it is an invariant that valid_len bytes are valid
            unsafe { core::str::from_utf8_unchecked(valid_bytes) }
        }
    }

    fn advance(&mut self, amount: u8) {
        debug_assert!(amount <= self.saved_len());

        self.lengths = (self.lengths & 0xF0) | ((self.saved_len() - amount) & 0xF);

        if self.valid_len() > 0 {
            debug_assert!(amount <= self.valid_len());
            self.lengths = ((self.valid_len() - amount) << 4) | (self.lengths & 0xF);
        }

        let start = usize::from(amount);
        let new_length = start + usize::from(self.saved_len());
        self.buffer.copy_within(start..new_length, 0);
    }

    fn take_char(&mut self) -> Result<Option<char>, NameError> {
        if self.valid_len() > 0 {
            let mut chars = self.valid().chars();
            let original_len = chars.as_str().len();

            let c = if cfg!(debug_assertions) {
                chars.next().unwrap()
            } else {
                // Safety: check above ensures string is not empty
                unsafe { chars.next().unwrap_unchecked() }
            };

            // Skip the amount of bytes that were read
            // will advance at most 15, which is less than u8::MAX
            #[allow(clippy::cast_possible_truncation)]
            self.advance((original_len - chars.as_str().len()) as u8);

            Ok(Some(c))
        } else if let Some(bad) = self.bad_sequence.take() {
            let bad_len = usize::from(bad.get());
            let mut bytes = [0u8; 4];
            bytes[0..bad_len].copy_from_slice(&self.buffer[0..bad_len]);

            // Skip the invalid sequence
            self.advance(bad.get());

            Err(NameError::BadBytes(InvalidCodePoint { length: bad, bytes }))
        } else {
            Ok(None)
        }
    }

    fn fill(
        &mut self,
        offset: &mut u64,
        length: &mut u32,
        input: &impl Input,
    ) -> Result<(), NameError> {
        // take_char would return Ok(Some) or Err for the same conditions here
        if self.valid_len() > 0 || self.bad_sequence.is_some() {
            return Ok(());
        }

        let saved_length = usize::from(self.saved_len());
        if saved_length < self.buffer.len() {
            let remaining = self.buffer.len() - saved_length;
            let actual_remaining = core::cmp::min(remaining, *length as usize);
            match input.read(offset, &mut self.buffer[saved_length..][..actual_remaining]) {
                Ok(amount) => {
                    let filled: u8;

                    #[allow(clippy::cast_possible_truncation)]
                    {
                        // buf.len() <= 15 <= u8::MAX
                        filled = amount as u8;

                        // saved_length <= 15 <= u8::MAX
                        self.lengths =
                            (self.lengths & 0xF0) | ((saved_length as u8 + filled) & 0xF);
                    }

                    *length -= u32::from(filled);
                }
                Err(e) => {
                    *length = 0;
                    self.lengths = 0;
                    self.bad_sequence = None;
                    return Err(NameError::BadInput(e));
                }
            }
        }

        debug_assert_eq!(self.valid_len(), 0);

        let valid_len = match core::str::from_utf8(&self.buffer[0..usize::from(self.saved_len())]) {
            Ok(s) => s.len(),
            Err(e) => {
                if let Some(bad_len) = e.error_len() {
                    self.bad_sequence = Some(
                        u8::try_from(bad_len)
                            .ok()
                            .and_then(core::num::NonZeroU8::new)
                            .unwrap(),
                    );
                }

                e.valid_up_to()
            }
        };

        #[allow(clippy::cast_possible_truncation)]
        {
            // valid_len <= 15 <= u8::MAX
            self.lengths = ((valid_len as u8) << 4) | (self.lengths & 0xF);
        }

        Ok(())
    }
}

/// An iterator over the [`char`]s of a [`Name`].
///
/// See the documentation for [`Name::chars()`] for more information.
#[derive(Clone, Copy)]
#[must_use]
pub struct Chars<I: Input> {
    name: Name<I>,
    buffer: CharsBuffer,
}

impl<I: Input> Chars<I> {
    pub(super) fn new(name: Name<I>) -> Self {
        Self {
            name,
            buffer: Default::default(),
        }
    }

    fn next_inner(&mut self) -> Result<Option<char>, NameError> {
        if self.name.length == 0 && self.buffer.saved_len() == 0 {
            return Ok(None);
        }

        self.buffer.fill(
            &mut self.name.offset,
            &mut self.name.length,
            &self.name.input,
        )?;

        self.buffer.take_char()
    }
}

impl<I: Input> HasInput<I> for Chars<I> {
    #[inline]
    fn input(&self) -> &I {
        self.name.input()
    }
}

impl<'a, I: Input + 'a> BorrowInput<'a, I> for Chars<I> {
    type Borrowed = Chars<&'a I>;

    #[inline]
    fn borrow_input(&'a self) -> Chars<&'a I> {
        Chars {
            name: self.name.borrow_input(),
            buffer: self.buffer,
        }
    }
}

impl<I: Input> Iterator for Chars<I> {
    type Item = Result<char, NameError>;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.next_inner().transpose()
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let max = usize::from(self.buffer.saved_len()) + (self.name.length as usize);
        (core::cmp::min(1, max), Some(max))
    }
}

impl<I: Input> core::iter::FusedIterator for Chars<I> {}

/// An iterator over the [`char`]s of a [`Name`] that substitutes invalid byte sequences and other errors with [`char::REPLACEMENT_CHARACTER`].
///
/// See the documentation for [`Name::chars_lossy()`] for more information.
#[derive(Clone, Copy)]
#[must_use]
pub struct CharsLossy<I: Input> {
    inner: Chars<I>,
}

impl<I: Input> CharsLossy<I> {
    #[inline]
    pub(super) fn new(inner: Chars<I>) -> CharsLossy<I> {
        Self { inner }
    }
}

impl<I: Input> HasInput<I> for CharsLossy<I> {
    #[inline]
    fn input(&self) -> &I {
        self.inner.input()
    }
}

impl<'a, I: Input + 'a> BorrowInput<'a, I> for CharsLossy<I> {
    type Borrowed = CharsLossy<&'a I>;

    #[inline]
    fn borrow_input(&'a self) -> Self::Borrowed {
        CharsLossy::new(self.inner.borrow_input())
    }
}

impl<I: Input> Iterator for CharsLossy<I> {
    type Item = char;

    fn next(&mut self) -> Option<char> {
        Some(match self.inner.next()? {
            Ok(c) => c,
            Err(_) => char::REPLACEMENT_CHARACTER,
        })
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.inner.size_hint()
    }
}

impl<I: Input> core::iter::FusedIterator for CharsLossy<I> {}
