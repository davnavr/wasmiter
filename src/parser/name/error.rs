use core::fmt::{Display, Formatter, Write as _};

/// Describes an invalid byte sequence or code point that was encountered while decoding a
/// [`Name`](crate::parser::name::Name).
#[derive(Clone, Copy)]
pub struct InvalidCodePoint {
    pub(super) length: core::num::NonZeroU8,
    pub(super) bytes: [u8; 4],
}

impl InvalidCodePoint {
    fn bytes(&self) -> Option<&[u8]> {
        let length = usize::from(self.length.get());
        if length <= self.bytes.len() {
            Some(&self.bytes[0..length])
        } else {
            None
        }
    }
}

impl core::fmt::Debug for InvalidCodePoint {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        if let Some(bytes) = self.bytes() {
            core::fmt::Debug::fmt(bytes, f)
        } else {
            f.debug_struct("InvalidCodePoint")
                .field("length", &self.length)
                .finish()
        }
    }
}

impl Display for InvalidCodePoint {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.write_str("encountered invalid byte sequence or invalid code point")?;
        if let Some(bytes) = self.bytes() {
            f.write_char(':')?;
            for b in bytes {
                write!(f, " {b:#04X}")?;
            }
            Ok(())
        } else {
            write!(f, " of length {}", self.length)
        }
    }
}

#[cfg(feature = "std")]
impl std::error::Error for InvalidCodePoint {}

/// Errors which can occur when attempting to interpret a [`Name`](crate::parser::name::Name) as a
/// UTF-8 string.
#[derive(Debug)]
pub enum NameError {
    /// An operation to read the UTF-8 string contents from the [`Bytes`](crate::bytes::Bytes)
    /// failed.
    BadInput(crate::bytes::Error),
    /// The UTF-8 string itself is malformed.
    BadBytes(InvalidCodePoint),
}

impl Display for NameError {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::BadInput(bad) => Display::fmt(bad, f),
            Self::BadBytes(e) => Display::fmt(e, f),
        }
    }
}

#[cfg(feature = "std")]
impl std::error::Error for NameError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::BadInput(bad) => Some(bad),
            Self::BadBytes(bad) => Some(bad),
        }
    }
}
