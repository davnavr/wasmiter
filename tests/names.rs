use wasmiter::parser::name::Name;

#[test]
fn basic_type() {
    let inputs = [
        "fd_write".as_bytes(),
        "wasi_snapshot_preview1".as_bytes(),
        "Hello World!".as_bytes(),
        &[0x61, 0x62, 0x63],
        &[0],
        "espa\u{F1}ol".as_bytes(),
        "\u{420}\u{43e}\u{441}\u{441}\u{438}\u{44f}".as_bytes(),
        "Fuerza A\u{E9}rea Mexicana".as_bytes(),
        "\u{1f643}".as_bytes(),
        &[0xFF, 0xFF], // Invalid
    ];

    insta::assert_debug_snapshot!(inputs
        .iter()
        .map(|bytes| Name::try_from(*bytes))
        .collect::<Box<[_]>>());
}
