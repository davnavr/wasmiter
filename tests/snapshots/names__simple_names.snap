---
source: tests/names.rs
expression: "inputs.iter().map(|bytes| Name::try_from(*bytes)).collect::<Box<[_]>>()"
---
[
    Ok(
        "fd_write",
    ),
    Ok(
        "wasi_snapshot_preview1",
    ),
    Ok(
        "Hello World!",
    ),
    Ok(
        "abc",
    ),
    Ok(
        "\0",
    ),
    Ok(
        "espa\u{0xF1}ol",
    ),
    Ok(
        "\u{0x420}\u{0x43E}\u{0x441}\u{0x441}\u{0x438}\u{0x44F}",
    ),
    Ok(
        "Fuerza A\u{0xE9}rea Mexicana",
    ),
    Ok(
        "\u{0x1F643}",
    ),
    Ok(
        "��",
    ),
]
